src/algebra/order/floor.lean

```diff
@@ -67,10 +63,19 @@ instance : floor_semiring ℕ :=
   gc_floor := λ n a ha, by { rw nat.cast_id, refl },
   gc_ceil := λ n a, by { rw nat.cast_id, refl } }
 
+@[priority 100] -- see Note [lower instance priority]
+instance ordered_semiring.nontrivial [ordered_semiring α] [floor_semiring α] : nontrivial α :=
+begin
+  casesI subsingleton_or_nontrivial α with _ h,
+  { refine false.elim (nat.not_succ_le_self (floor_semiring.floor (0 : α)) _),
+    exact (floor_semiring.gc_floor le_rfl).2 (subsingleton.le _ _) },
+  { exact h }
```

Comment:
```suggestion
  refine (subsingleton_or_nontrivial α).resolve_left (λ h, _),
  refine false.elim (nat.not_succ_le_self (floor_semiring.floor (0 : α)) _),
  exact (floor_semiring.gc_floor le_rfl).2 (@@subsingleton.le _ h _ _),
```

a little golf - the same style is possible on the other lemma too. also you could use `by exactI` instead of `@@... _ h`